Mathlib/GroupTheory/GroupAction/Period.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2024 Emilie Burgun. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Emilie Burgun
+-/
+
+import Mathlib.GroupTheory.GroupAction.Basic
+import Mathlib.Dynamics.PeriodicPts
+import Mathlib.GroupTheory.Exponent
+
+/-!
+# Period of a group action
+
+This module defines some helpful lemmas around [`MulAction.period`] and [`AddAction.period`].
+The period of a point `a` by a group element `g` is the smallest `m` such that `g ^ m • a = a`
+(resp. `(m • g) +ᵥ a = a`) for a given `g : G` and `a : α`.
+
+If such an `m` does not exist,
+then by convention `MulAction.period` and `AddAction.period` return 0.
+-/
+
+namespace MulAction
+
+universe u v
+variable {α : Type v}
+variable {G : Type u} [Group G] [MulAction G α]
+variable {M : Type u} [Monoid M] [MulAction M α]
+
+/-- If the action is periodic, then a lower bound for its period can be computed. -/
+@[to_additive]
+theorem period_gt_of_moved {m : M} {a : α} {n : ℕ} (period_pos : 0 < period m a)
+    (moved : ∀ k, 0 < k → k < n → m^k • a ≠ a) : n ≤ period m a := by
+  by_contra period_le_n
+  rw [not_le] at period_le_n
+  apply moved _ period_pos period_le_n
+  exact smul_pow_period_fixed m a
```

Comment:
`by_contra!` can do some of the rewrites for you automatically:
```suggestion
  by_contra! period_lt_n
  apply moved _ period_pos period_lt_n
  exact smul_pow_period_fixed m a
```

Reply:
In fact if you really love term mode this could even become:
```suggestion
  le_of_not_gt <| fun period_lt_n => moved _ period_pos period_lt_n <| smul_pow_period_fixed m a
```
(although I would not advise it!)